Util/Tokenization/src/main/java/org/tribuo/util/tokens/impl/wordpiece/WordpieceTokenizer.java

```diff
@@ -1,31 +1,67 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.tribuo.util.tokens.impl.wordpiece;
 
 import java.text.Normalizer;
 import java.util.ArrayList;
-import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
 
 import org.tribuo.util.tokens.Token;
 import org.tribuo.util.tokens.Token.TokenType;
 import org.tribuo.util.tokens.Tokenizer;
+import org.tribuo.util.tokens.impl.WhitespaceTokenizer;
 
 import com.oracle.labs.mlrg.olcut.config.Config;
 import com.oracle.labs.mlrg.olcut.provenance.ConfiguredObjectProvenance;
 import com.oracle.labs.mlrg.olcut.provenance.impl.ConfiguredObjectProvenanceImpl;
 
+/**
+ * This Tokenizer is meant to be a reasonable approximation of the BertTokenizer
+ * defined <a href=
+ * "https://github.com/huggingface/transformers/blob/master/src/transformers/models/bert/tokenization_bert.py#L117">here</a>.
+ * Please see class definition for <code>BertTokenizer</code> (the line numbers
+ * may change.) Please see notes in WordpieceTokenizerTest for information about
+ * how we tested the similarity between this tokenizer and the referenced python
+ * implementation and for regression test examples that fail. In short, there
+ * are outstanding discrepancies for texts that include Arabic and other
+ * non-latin scripts that generate so many "[UNK]" tokens for an English-based
+ * BPE vocabulary as to render the discrepancies as practically meaningless.
+ * <p>
+ * As in the reference implementation, the input text is whitespace tokenized
+ * and each token is further tokenized to account for things like punctuation
+ * and Chinese characters. The resulting tokens are then applied to the
+ * wordpiece algorithm implemented in {@link Wordpiece} which is driven by an
+ * input vocabulary which matches tokens and token suffixes as it can. Any
+ * tokens that are not found in the input vocbulary are marked as "unknown".
+ */
 public class WordpieceTokenizer implements Tokenizer {
 
-    @Config
+    @Config(mandatory=true, description="an instance of Wordpiece which applies the 'wordpiece' algorithm")
     private Wordpiece wordpiece;
-    @Config
+    @Config(description="determines whether or not to lowercase the input text")
     private boolean toLowerCase = true;
-    @Config
-    private Tokenizer tokenizer = new WordpiecePreprocessTokenizer();
-    @Config
+    @Config(description="performs whitespace tokenization before 'basic' tokenizer is applied (see next)")
```

Comment:
The ordering of configuration fields when we print out the descriptions is controlled by the JVM and there's no guarantee it's in source file order.

Reply:
got it.  updated to "see basicTokenizer"